src/libstd/thread/block_on_future.rs

```diff
@@ -0,0 +1,236 @@
+//! Future/Task/Thread integration.
+//! The method defined in this module allows to block a thread until an
+//! async task completes.
+
+use crate::future::Future;
+use crate::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+use crate::mem;
+use crate::pin::Pin;
+use crate::sync::Arc;
+use super::{current, park, Inner, Thread};
+
+/// Carries a flag that is used to wakeup the executor.
+/// A pointer to this struct is passed to the thread-local waker.
+struct LocalWakeState {
+    is_woken: bool,
+    waker_was_cloned: bool,
+}
+
+/// Returns the vtable that is used for waking up the executor
+/// from another thread.
+fn threadsafe_waker_vtable() -> &'static RawWakerVTable {
+    &RawWakerVTable::new(
+        clone_threadsafe_waker,
+        wake_threadsafe_waker,
+        wake_threadsafe_waker_by_ref,
+        drop_threadsafe_waker,
+    )
+}
+
+/// Returns the vtable that is used for waking up the executor
+/// from inside it's execution on the current thread.
+fn current_thread_waker_vtable() -> &'static RawWakerVTable {
+    &RawWakerVTable::new(
+        create_threadsafe_waker,
+        wake_current_thread,
+        wake_current_thread_by_ref,
+        |_| {},
+    )
+}
+
+/// This method will be called when the waker reference gets cloned,
+/// which makes it possible to transfer it to another thread. In this
+/// case we have to create a threadsafe `Waker`. In order to to this
+/// we retain the thread handle and store it in the new `RawWaker`s
+/// data pointer.
+unsafe fn create_threadsafe_waker(data: *const()) -> RawWaker {
+    let wake_state = data as *mut LocalWakeState;
+    (*wake_state).waker_was_cloned = true;
+
+    // Get the `Arc<Inner>` of a current thread handle and store into in
+    // the type erased pointer.
+    //
+    // This mechanism exploits the fact that `Thread` is already an `Arc<Inner>`.
+    // Therefore in order to create clones of the thread handle we can just
+    // directly create clones of the Inner state, and recreate the `Thread`
+    // handle when necessary.
+    //
+    // If the implementation of `Thread` would be changed to something different,
+    // we would need to wrap the complete `Thread` object in another `Arc` by
+    // adopt the following line to:
+    // `let arc_thread = Arc::new(current());`
+    let arc_thread_inner = current().inner;
```

Comment:
Both &Context and &Waker are Send, so they might be send to another thread
(however unlikely it is to happen in practice). Thus it isn't a valid
optimization to delay obtaining the current thread.

Reply:
Oh, that's a bummer. I somehow missed that `Waker` is `Sync` - didn't remember we made it that way. Guess I will remove the optimization again. 

`LocalWaker`  would also have been nice for this use-case, but that's also gone.